File: lib/circuit_breaker/worker.ex
defmodule CircuitBreaker.Worker do
  use GenServer

  alias CircuitBreaker.{Main, Storage}

  def start_link(_opts), do: GenServer.start_link(__MODULE__, %{}, name: __MODULE__)

  @impl true
  def init(state) do
    Storage.repair_running_state()
    release_finished_run_reservations()
    send(self(), :process)
    {:ok, state}
  end

  @impl true
  def handle_info(:process, state) do
    process_once()
    Process.send_after(self(), :process, 10)
    {:noreply, state}
  end

  def run_forever do
    Storage.repair_running_state()
    release_finished_run_reservations()
    loop_forever()
  end

  def process_once(poll_timeout \\ 1), do: process_once_internal(poll_timeout, true)
  def process_once_sync(poll_timeout \\ 1), do: process_once_internal(poll_timeout, false)

  defp process_once_internal(poll_timeout, async?) do
    case Storage.pop_task_id(poll_timeout) do
      nil ->
        false

      task_id ->
        task = Storage.get_task(task_id)

        cond do
          is_nil(task) or task["status"] in ~w(running completed failed) ->
            false

          true ->
            run = Storage.get_run(task["run_id"])

            if run && can_start_task?(run, task) do
              if async? do
                Task.start(fn -> execute_task(run, task) end)
              else
                execute_task(run, task)
              end

              true
            else
              refreshed = Storage.get_task(task_id)

              if refreshed && refreshed["status"] not in ~w(completed failed) do
                Storage.enqueue_task(task_id)
              end

              Process.sleep(200)
              false
            end
        end
    end
  end

  def run_subprocess_simulation(task) do
    outcome = if task["planned_first_attempt_outcome"], do: outcome_for_attempt(task)

    duration =
      if to_int(Map.get(task, "attempt", 1)) > 1 && task["retry_duration"] != nil do
        to_float(task["retry_duration"])
      else
        to_float(task["duration"])
      end

    scale = Application.get_env(:circuit_breaker, :task_duration_scale, 1.0)
    timeout = Main.max_subprocess_timeout_seconds()

    cond do
      outcome == "timeout" ->
        sleep_seconds(max(duration, timeout) * scale)
        raise TimeoutError, message: "subprocess reached #{timeout} second kill timeout"

      duration * scale >= timeout ->
        sleep_seconds(timeout)
        raise TimeoutError, message: "subprocess reached #{timeout} second kill timeout"

      true ->
        sleep_seconds(duration * scale)
        :ok
    end
  end

  def delay_before_retry(failed_count) do
    delay = Main.retry_delay_seconds(failed_count) * Application.get_env(:circuit_breaker, :task_duration_scale, 1.0)
    if delay > 0, do: sleep_seconds(delay)
  end

  defp can_start_task?(run, task) do
    summary = Storage.summarize_running_tasks(run["run_id"])

    cond do
      summary["global_running"] >= summary["global_limit"] ->
        task
        |> Map.put("blocked_reason", "global_concurrency_limit")
        |> Map.put("message", "Task is waiting because the global subtask process limit of #{summary["global_limit"]} running processes has been reached.")
        |> Storage.update_task()

        false

      summary["run_running"] >= to_int(run["max_concurrency"]) ->
        task
        |> Map.put("blocked_reason", "run_concurrency_limit")
        |> Map.put("message", "Task is waiting because the run concurrency limit of #{run["max_concurrency"]} running tasks has been reached.")
        |> Storage.update_task()

        false

      not Storage.acquire_slots(run["run_id"], to_int(run["max_concurrency"])) ->
        task
        |> Map.put("blocked_reason", "global_concurrency_limit")
        |> Map.put("message", "Task is waiting for an available execution slot.")
        |> Storage.update_task()

        false

      not Main.memory_is_available(Main.min_available_memory_mb()) ->
        Storage.release_slots(run["run_id"])

        task
        |> Map.put("blocked_reason", "insufficient_memory")
        |> Map.put("message", "Task is waiting because available memory is below #{Main.min_available_memory_mb()} MB.")
        |> Storage.update_task()

        false

      true ->
        true
    end
  end

  defp execute_task(run, task) do
    task =
      task
      |> Map.put("blocked_reason", nil)
      |> Map.put("status", "running")
      |> Map.put("attempt", to_int(task["attempt"]) + 1)
      |> Map.put("started_at", task["started_at"] || Storage.utc_now())
      |> Map.put("message", "Task attempt #{to_int(task["attempt"]) + 1} is running.")

    Storage.update_task(task)
    Storage.mark_run_running(run["run_id"])

    try do
      run_subprocess_simulation(task)
      handle_planned_outcome(task)
    rescue
      _ in TimeoutError ->
        handle_timeout(task)

      exc ->
        handle_unknown(task, exc)
    after
      Storage.release_slots(run["run_id"])
      Storage.refresh_run_terminal_state(run["run_id"])
      release_run_reservation_if_terminal(run["run_id"])
    end
  end

  defp handle_planned_outcome(task) do
    outcome = outcome_for_attempt(task)

    if outcome == "completed" do
      task
      |> Map.put("status", "completed")
      |> Map.put("result", task["planned_result"])
      |> Map.put("error", nil)
      |> Map.put("message", "Task completed successfully.")
      |> Map.put("finished_at", Storage.utc_now())
      |> Storage.update_task()
    else
      task = task |> Map.put("error", error_for_outcome(outcome)) |> Map.put("reason", outcome)

      if can_retry?(task) do
        task
        |> Map.put("status", "retrying")
        |> Map.put("message", "Task attempt #{task["attempt"]} ended with #{outcome}; retrying.")
        |> Storage.update_task()

        delay_before_retry(to_int(task["attempt"]))
        Storage.enqueue_retry_task(task["task_id"])
      else
        task
        |> Map.put("status", "failed")
        |> Map.put("message", "Task failed after #{task["attempt"]} attempts. Last reason: #{outcome}.")
        |> Map.put("finished_at", Storage.utc_now())
        |> Storage.update_task()
      end
    end
  end

  defp handle_timeout(task) do
    task = task |> Map.put("error", "task_timeout") |> Map.put("reason", "timeout")

    if can_retry?(task) do
      task
      |> Map.put("status", "retrying")
      |> Map.put("message", "Task attempt #{task["attempt"]} reached the #{Main.max_subprocess_timeout_seconds()} second kill timeout; retrying.")
      |> Storage.update_task()

      delay_before_retry(to_int(task["attempt"]))
      Storage.enqueue_retry_task(task["task_id"])
    else
      task
      |> Map.put("status", "failed")
      |> Map.put("message", "Task failed because the subprocess reached the #{Main.max_subprocess_timeout_seconds()} second kill timeout twice.")
      |> Map.put("finished_at", Storage.utc_now())
      |> Storage.update_task()
    end
  end

  defp handle_unknown(task, exc) do
    task =
      task
      |> Map.put("error", "unknown_error")
      |> Map.put("reason", inspect(exc.__struct__))
      |> Map.put("message", "Task failed with unexpected error: #{Exception.message(exc)}")

    if can_retry?(task) do
      task |> Map.put("status", "retrying") |> Storage.update_task()
      delay_before_retry(to_int(task["attempt"]))
      Storage.enqueue_retry_task(task["task_id"])
    else
      task |> Map.put("status", "failed") |> Map.put("finished_at", Storage.utc_now()) |> Storage.update_task()
    end
  end

  defp can_retry?(task), do: to_int(task["attempt"]) <= Main.max_task_retries()

  defp outcome_for_attempt(task) do
    if to_int(task["attempt"]) == 1 do
      task["planned_first_attempt_outcome"]
    else
      task["planned_retry_attempt_outcome"] || task["planned_first_attempt_outcome"]
    end
  end

  defp error_for_outcome("failed"), do: "task_failed"
  defp error_for_outcome("timeout"), do: "task_timeout"
  defp error_for_outcome(_), do: "unknown_error"

  defp release_run_reservation_if_terminal(run_id) do
    if amount = Storage.mark_run_concurrency_released(run_id) do
      Main.release_run_concurrency(amount)
    end
  end

  defp release_finished_run_reservations do
    Redix.command!(Storage.redis_name(), ["SCAN", "0", "MATCH", "run:*"])
    |> scan_release_keys()
  end

  defp scan_release_keys([cursor, keys]) do
    keys
    |> Enum.reject(&(String.ends_with?(&1, ":tasks") or String.ends_with?(&1, ":running_tasks")))
    |> Enum.each(fn "run:" <> run_id -> release_run_reservation_if_terminal(run_id) end)

    if cursor != "0" do
      Redix.command!(Storage.redis_name(), ["SCAN", cursor, "MATCH", "run:*"])
      |> scan_release_keys()
    end
  end

  defp loop_forever do
    process_once()
    loop_forever()
  end

  defp sleep_seconds(seconds), do: Process.sleep(round(seconds * 1000))
  defp to_int(value) when is_integer(value), do: value
  defp to_int(value) when is_binary(value), do: String.to_integer(value)
  defp to_float(value) when is_float(value), do: value
  defp to_float(value) when is_integer(value), do: value * 1.0
  defp to_float(value) when is_binary(value), do: String.to_float(value)
end

defmodule TimeoutError do
  defexception [:message]
end

File: lib/circuit_breaker/worker.rb
# frozen_string_literal: true

module CircuitBreaker
  class TaskTimeout < StandardError; end

  module Worker
    module_function

    def run_forever
      Storage.repair_running_state
      release_finished_run_reservations
      loop do
        process_once
      end
    end

    def process_once(poll_timeout = 1, async: false)
      task_id = Storage.pop_task_id(poll_timeout)
      return false unless task_id

      task = Storage.get_task(task_id)
      return false if task.nil? || %w[running completed failed].include?(task["status"])

      run = Storage.get_run(task["run_id"])
      if run && can_start_task?(run, task)
        if async
          Thread.new { execute_task(run, task) }
        else
          execute_task(run, task)
        end
        true
      else
        refreshed = Storage.get_task(task_id)
        Storage.enqueue_task(task_id) if refreshed && !%w[completed failed].include?(refreshed["status"])
        sleep 0.2
        false
      end
    end

    def run_subprocess_simulation(task)
      outcome = outcome_for_attempt(task)
      duration = if task["attempt"].to_i > 1 && !task["retry_duration"].nil?
        task["retry_duration"].to_f
      else
        task["duration"].to_f
      end
      scale = Main.task_duration_scale
      timeout = Main.max_subprocess_timeout_seconds

      if outcome == "timeout"
        sleep_seconds([duration, timeout].max * scale)
        raise TaskTimeout, "subprocess reached #{timeout} second kill timeout"
      elsif duration * scale >= timeout
        sleep_seconds(timeout)
        raise TaskTimeout, "subprocess reached #{timeout} second kill timeout"
      else
        sleep_seconds(duration * scale)
      end
    end

    def delay_before_retry(failed_count)
      delay = Main.retry_delay_seconds(failed_count) * Main.task_duration_scale
      sleep_seconds(delay) if delay.positive?
    end

    def can_start_task?(run, task)
      summary = Storage.summarize_running_tasks(run["run_id"])

      if summary["global_running"] >= summary["global_limit"]
        Storage.update_task(task.merge(
          "blocked_reason" => "global_concurrency_limit",
          "message" => "Task is waiting because the global subtask process limit of #{summary["global_limit"]} running processes has been reached."
        ))
        return false
      end

      if summary["run_running"] >= run["max_concurrency"].to_i
        Storage.update_task(task.merge(
          "blocked_reason" => "run_concurrency_limit",
          "message" => "Task is waiting because the run concurrency limit of #{run["max_concurrency"]} running tasks has been reached."
        ))
        return false
      end

      unless Storage.acquire_slots(run["run_id"], run["max_concurrency"].to_i)
        Storage.update_task(task.merge(
          "blocked_reason" => "global_concurrency_limit",
          "message" => "Task is waiting for an available execution slot."
        ))
        return false
      end

      unless Main.memory_is_available(Main.min_available_memory_mb)
        Storage.release_slots(run["run_id"])
        Storage.update_task(task.merge(
          "blocked_reason" => "insufficient_memory",
          "message" => "Task is waiting because available memory is below #{Main.min_available_memory_mb} MB."
        ))
        return false
      end

      true
    end

    def execute_task(run, task)
      attempt = task["attempt"].to_i + 1
      task = task.merge(
        "blocked_reason" => nil,
        "status" => "running",
        "attempt" => attempt,
        "started_at" => task["started_at"] || Storage.utc_now,
        "message" => "Task attempt #{attempt} is running."
      )
      Storage.update_task(task)
      Storage.mark_run_running(run["run_id"])

      begin
        run_subprocess_simulation(task)
        handle_planned_outcome(task)
      rescue TaskTimeout
        handle_timeout(task)
      rescue StandardError => e
        handle_unknown(task, e)
      ensure
        Storage.release_slots(run["run_id"])
        Storage.refresh_run_terminal_state(run["run_id"])
        release_run_reservation_if_terminal(run["run_id"])
      end
    end

    def handle_planned_outcome(task)
      outcome = outcome_for_attempt(task)
      if outcome == "completed"
        Storage.update_task(task.merge(
          "status" => "completed",
          "result" => task["planned_result"],
          "error" => nil,
          "message" => "Task completed successfully.",
          "finished_at" => Storage.utc_now
        ))
      else
        task = task.merge("error" => error_for_outcome(outcome), "reason" => outcome)
        if can_retry?(task)
          Storage.update_task(task.merge("status" => "retrying", "message" => "Task attempt #{task["attempt"]} ended with #{outcome}; retrying."))
          delay_before_retry(task["attempt"].to_i)
          Storage.enqueue_retry_task(task["task_id"])
        else
          Storage.update_task(task.merge(
            "status" => "failed",
            "message" => "Task failed after #{task["attempt"]} attempts. Last reason: #{outcome}.",
            "finished_at" => Storage.utc_now
          ))
        end
      end
    end

    def handle_timeout(task)
      task = task.merge("error" => "task_timeout", "reason" => "timeout")
      if can_retry?(task)
        Storage.update_task(task.merge(
          "status" => "retrying",
          "message" => "Task attempt #{task["attempt"]} reached the #{Main.max_subprocess_timeout_seconds} second kill timeout; retrying."
        ))
        delay_before_retry(task["attempt"].to_i)
        Storage.enqueue_retry_task(task["task_id"])
      else
        Storage.update_task(task.merge(
          "status" => "failed",
          "message" => "Task failed because the subprocess reached the #{Main.max_subprocess_timeout_seconds} second kill timeout twice.",
          "finished_at" => Storage.utc_now
        ))
      end
    end

    def handle_unknown(task, error)
      task = task.merge(
        "error" => "unknown_error",
        "reason" => error.class.name,
        "message" => "Task failed with unexpected error: #{error.message}"
      )

      if can_retry?(task)
        Storage.update_task(task.merge("status" => "retrying"))
        delay_before_retry(task["attempt"].to_i)
        Storage.enqueue_retry_task(task["task_id"])
      else
        Storage.update_task(task.merge("status" => "failed", "finished_at" => Storage.utc_now))
      end
    end

    def can_retry?(task)
      task["attempt"].to_i <= Main.max_task_retries
    end

    def outcome_for_attempt(task)
      task["attempt"].to_i == 1 ? task["planned_first_attempt_outcome"] : (task["planned_retry_attempt_outcome"] || task["planned_first_attempt_outcome"])
    end

    def error_for_outcome(outcome)
      { "failed" => "task_failed", "timeout" => "task_timeout" }.fetch(outcome, "unknown_error")
    end

    def release_run_reservation_if_terminal(run_id)
      amount = Storage.mark_run_concurrency_released(run_id)
      Main.release_run_concurrency(amount) if amount
    end

    def release_finished_run_reservations
      Storage.scan_keys("run:*").reject { |key| key.end_with?(":tasks", ":running_tasks") }.each do |key|
        release_run_reservation_if_terminal(key.delete_prefix("run:"))
      end
    end

    def sleep_seconds(seconds)
      sleep(seconds)
    end
  end
end
